[클린코드 7기 김진주] 영화 리뷰 미션 STEP 3

devpearlkim opened the pull request:
안녕하세요 윤호님!
이번 pr은 사실, 테스트코드 작성하고 코드를 다듬은 뒤에 올려야 하는데,
아무래도 요즘 시간이 없어서 기한을 채우지 못할 것 같아 그냥 올립니다.

[깃헙 페이지](https://devpearlkim.github.io/js-movie-review/)

피드백을 주셔도 바로 반영할 수 있을지는 모르겠지만,
어렵다면 나중에라도 혼자 수정해보겠습니다.

윤호님께서 처음에 같은 레벨의 코드 일관성에 대해 말씀해 주셨는데,
그런 개념들을 처음 배우고 실제로 일하며 집중해서 적용해봤습니다.

집중해서 쓰다 보니 왜 같은 레벨에 있는 컴포넌트는 비슷한 형태와 책임을 가져야 하는지,
왜 하나의 함수가 하나의 역할만 해야하는지,
도메인 데이터와 UI 데이터 구조를 분리하는게 어떻게 비즈니스로직과 UI로직을 분리하는데 영향을 미치는지,
그런 것들에 대해 생각해볼 수 있었습니다.

아직은 열 번 코드를 작성하면 한 번 정도만 제대로 적용하는 수준이지만,
계속 연습해서 자연스럽게 써먹을 수 있도록 노력하겠습니다.

그간 감사했습니다.

devpearlkim committed:
feat: Modularized HTML structure

devpearlkim committed:
feat: Fetch and display movie list from TMDB API

devpearlkim committed:
feat: Separate header, footer, and tabs components & add skeleton UI for movie list

devpearlkim committed:
feat: Implement "Load More" functionality and modularize movie list components

- Added "Load More" button to dynamically load additional movies
- Separated movie rendering logic into `movieRenderer.js`
- Moved "Load More" button creation/removal to `loadMoreButton.js`

devpearlkim committed:
chore: Delete custom button style

devpearlkim committed:
feat: Add responsive design & header background image

devpearlkim committed:
feat: Dynamically update header with first movie & add error message UI

devpearlkim committed:
chore: ensure previous movies persist when loading more

devpearlkim committed:
feat: Separate error handling and UI feedback for better user experience

devpearlkim committed:
chore: Setup Cypress for E2E testing

devpearlkim committed:
refactor: Improve Tabs component, add Header template, and optimize resize event

- Replaced innerHTML in Tabs with createElement for better security and performance
- Converted Header to use a template for cleaner separation of structure and logic
- Added debounce to resize event to optimize performance

devpearlkim committed:
refactor: Optimize DOM updates in movieRenderer.js using DocumentFragment

devpearlkim committed:
feat: Update movie fetching based on selected tab

devpearlkim committed:
test: Add skeleton UI and tab functionality tests

devpearlkim committed:
test: Add tests for data fetching and load more functionality

devpearlkim committed:
test: Add header rendering and error handling tests

devpearlkim committed:
chore: Delete unused file

devpearlkim committed:
refactor: Separate domain logic and UI with functional approach

devpearlkim committed:
feat: Add search functionality with URL state management

- Separated header logic into LoadBaseHeader (renders base header with search form) and updateHeaderMovie (updates movie info)
- Added search service logic to fetch search results via API
- Updated controller to attach search form submit event and handle search actions

devpearlkim committed:
refactor: Reorganize controller for mode management and event handling

- Moved base header loading to main for one-time initialization
- Modified controller to attach search form event only once
- Implemented mode state management ("search" vs "category") with separate functions (loadCategory, handleSearch)
- Added popstate handling for proper history state restoration
- Updated tabs to only trigger switchTab in the controller

devpearlkim committed:
refactor: migrate project to Typescript

devpearlkim committed:
fix: STEP1 피드백 반영

devpearlkim committed:
test: Switch error UI test to alert-based error messages

devpearlkim committed:
refactor: Replace DOM manipulation with template literals for better readability

devpearlkim committed:
test: Add E2E tests for search functionality and query handling

devpearlkim committed:
chore: Fix import path of constants

devpearlkim committed:
chore: remove unused variable and api.js

devpearlkim committed:
refactor: Abstract common fetch logic into reusable fetchData helper

devpearlkim committed:
refactor: Update naming conventions for movie-relatited types

devpearlkim committed:
fix: Ensure non-null container assignment

devpearlkim committed:
refactor: Replace setMode with updateTabContainer and resetSearchInput function

devpearlkim committed:
refactor: Improve modularity and state management of MovieController

- Extract UI-related functions to `ui.ts`
- Introduce `currentMode` for better state handling
- Refactor `fetchMovies` logic for improved readability

devpearlkim committed:
refactor: Extract getCurrentMode

devpearlkim committed:
chore: Handle potential null return from createMovieController

devpearlkim committed:
refactor: Remove currentCategory and pass Tabs to controller init

devpearlkim committed:
chore: Move tabItems, validCategories, and defaultCategory to constants

devpearlkim committed:
refactor: Replace 'Load More' button with inifinite scroll

devpearlkim committed:
fix: Preserve existing movie list when loading new movies

devpearlkim committed:
refactor: Optimize renderNextBatch by reusing fetchMovies to eliminate redundancy

devpearlkim committed:
feat: Add movie details modal with API Integration

- Implement fetchMovieDetails API to fetch movie details
- Add getMovieDetails method in the service to return a MovieModel
- Connect movie click event in init() to fetchAndShowMovieDetails
- Fetch movie details on click and pass data to modal.open()
- Create Modal.ts with render, open, and close functions
- Ensure modal opens with formatted movie details and closes on ESC or background click

devpearlkim committed:
chore: Updated MovieApiDto and MovieModel

- Add release date, genres, and overview fields

devpearlkim committed:
feat: Add custom rating feature with localStorage persistence

devpearlkim committed:
chore: Delete unused function

devpearlkim committed:
docs: Update Requirements

devpearlkim committed:
refactor: Convert functions to arrow functions for consistency

devpearlkim committed:
fix: Prevent duplicate history entries for identical URLs

devpearlkim committed:
feat: Dynamically update section title based on search and category

devpearlkim committed:
Merge remote-tracking branch 'upstream/devpearlkim' into step3

yujo11 commented on src/utils/ui.ts:
```diff
@@ -60,4 +31,25 @@ export function updateHeader(service: IMovieService) {
+  queryOrCategory?: string
+): void {
+  const headerTitleEl = document.querySelector("main h2");
+  console.log(headerTitleEl);
```
👀

yujo11 commented on src/utils/ui.ts:
```diff
@@ -60,4 +31,25 @@ export function updateHeader(service: IMovieService) {
+  mode: "search" | "category",
+  queryOrCategory?: string
+): void {
+  const headerTitleEl = document.querySelector("main h2");
```
querySelector로 이렇게 트리를 타고 가서 DOM 요소를 파싱하는 경우 HTML의 구조 변경에 취약한데요.

`id`나 `class`를 적극적으로 활용해보시면 좋을거 같습니다 !

devpearlkim replied:
앗 넵 감사합니다!

yujo11 commented on src/utils/storage.ts:
Storage에 접근하는 함수를 따로 분리해주셨군요 좋네용 👍 👍 👍

yujo11 commented on src/utils/storage.ts:
```diff
@@ -0,0 +1,27 @@
+import { STORAGE_KEYS } from "../constants";
+import { UserRating } from "../types/type";
+
+export const getUserRating = (movieId: number): UserRating | null => {
+  const storedRatings: Record<number, UserRating> = JSON.parse(
+    localStorage.getItem(STORAGE_KEYS.USER_RATINGS) || "{}"
+  );
+
+  return storedRatings[movieId] || null;
```
일반적으로 `utils/` 에는 도메인과 무관하게 사용할 수 있는 함수들을 위치 시키는데요. 여기에서 선언 된 함수들이 이미 domain에 대한 정보를 알고 있어 조금 어색하게 느껴지는거 같아요

devpearlkim replied:
넵 감사합니다, util 전반적으로 도메인 정보를 알고 있는 부분이 많네요, 앞으로 유의하겠습니다

yujo11 commented on src/utils/state.ts:
```diff
@@ -1,6 +1,6 @@
 import { resetSearchInput, setSearchInput, updateTabContainer } from "./ui";
 
-export function getCurrentMode(query?: string): "search" | "category" {
+export const getCurrentMode = (query?: string): "search" | "category" => {
```
여기는 코드베이스 내 일관성을 지키기 위해 화살표 함수로 변경을 해주신걸까요?

devpearlkim replied:
넵 맞습니다,,!

yujo11 commented on src/utils/infiniteScroll.ts:
```diff
@@ -0,0 +1,35 @@
+const initInfiniteScroll = (options: InfiniteScrollOptions) => {
+  const { container, onLoadMore, hasMore, threshold = 1 } = options;
+
+  const observer = new IntersectionObserver(
```
👍 👍 👍

yujo11 commented on src/utils/infiniteScroll.ts:
```diff
@@ -0,0 +1,35 @@
+export interface InfiniteScrollOptions {
+  container: HTMLElement;
+  onLoadMore: () => void;
+  hasMore: () => void;
```
`hasMore`가 boolean을 반환할 것처럼 생겼네용

yujo11 commented on src/utils/infiniteScroll.ts:
```diff
@@ -0,0 +1,35 @@
+
+  const observer = new IntersectionObserver(
+    (entries) => {
+      if (entries[0].isIntersecting && hasMore()) {
```
여기 사용되는 곳을 보면 실제로 boolean을 반환할거 같기도 한데 맞나용..?

devpearlkim replied:
앗 넵,,! boolean을 반환하는데 훅에서 타입을 잘못 적었습니다, 감사합니다

yujo11 commented on src/components/Modal.ts:
```diff
@@ -0,0 +1,140 @@
+    return template.firstElementChild as HTMLElement;
+  };
+
+  const render = () => {
+    if (!modalElement) {
+      modalElement = createModalTemplate();
+      document.body.appendChild(modalElement);
+
+      starsContainer = modalElement.querySelector(".stars");
+      ratingText = modalElement.querySelector(".user-rating-text");
+
+      modalElement
+        .querySelector(".close-modal")
+        ?.addEventListener("click", close);
+      modalElement.addEventListener("click", (e) => {
+        if (e.target === modalElement) close();
+      });
+      document.addEventListener("keydown", (e) => {
+        if (e.key === "Escape") close();
+      });
+
+      starsContainer?.addEventListener("mouseover", hoverStarRating);
+      starsContainer?.addEventListener("mouseleave", resetStarRating);
+      starsContainer?.addEventListener("click", setUserRatingClick);
+    }
+  };
```
early return을 사용해서 depth를 줄여보셔도 좋겠네용 ㅎㅎ
```suggestion
  const render = () => {
    if (modalElement) return;

      modalElement = createModalTemplate();
      document.body.appendChild(modalElement);

      starsContainer = modalElement.querySelector(".stars");
      ratingText = modalElement.querySelector(".user-rating-text");

      modalElement
        .querySelector(".close-modal")
        ?.addEventListener("click", close);
      modalElement.addEventListener("click", (e) => {
        if (e.target === modalElement) close();
      });
      document.addEventListener("keydown", (e) => {
        if (e.key === "Escape") close();
      });

      starsContainer?.addEventListener("mouseover", hoverStarRating);
      starsContainer?.addEventListener("mouseleave", resetStarRating);
      starsContainer?.addEventListener("click", setUserRatingClick);

  };
```

devpearlkim replied:
넵,,! 감사합니다

yujo11 commented on src/components/Modal.ts:
```diff
@@ -0,0 +1,140 @@
+        if (e.key === "Escape") close();
+      });
+
+      starsContainer?.addEventListener("mouseover", hoverStarRating);
+      starsContainer?.addEventListener("mouseleave", resetStarRating);
```
오 hover 시 효과까지 구현해주셨군요 짱이네요 👍 👍 👍

yujo11 commented on src/components/Modal.ts:
```diff
@@ -0,0 +1,140 @@
+      modalElement.addEventListener("click", (e) => {
+        if (e.target === modalElement) close();
+      });
+      document.addEventListener("keydown", (e) => {
```
`keydown`과 `keyup`은 어떤 차이가 있을까요?

devpearlkim replied:
keydown은 키보드를 누르는 순간, keyup은 키보드를 눌렀다가 떼는 순간 이벤트가 발생한다고 나오는데,
사실 마우스나 키이벤트 쓸 때 어떤 이벤트가 적절한지 지금껏 고민해본적은 없었던 것 같습니다.
이 부분에 대해 알고 사용해야겠다는 생각이 드네요 감사합니다,,!

yujo11 replied:
추가로 `keydown`의 경우 키를 계속해서 누르고 있으면 연속해서 이벤트가 트리거 되는데요. 이 부분 때문에 연속적인 이벤트 트리거를 막기 위해 `keyup`을 쓰는 경우도 있으니 참고하시면 좋겠네용 ㅎㅎ (지금 변경하라는 얘기는 아닙니다!)

yujo11 commented on src/components/Modal.ts:
```diff
@@ -0,0 +1,140 @@
+    starsContainer.innerHTML = "";
+    starsContainer.appendChild(fragment);
+
+    ratingText!.textContent = RATING_MAP[score]
```
실제로 `ratingText`가 `null`인 경우에 어떤 동작이 발생할까요?

devpearlkim replied:
타입에러를 편하게 안보이게 하려고 사용했는데, null이 될 수도 있을 것 같습니다.
null일 경우에 대한 처리를 지금 시점에서도 못했는데,
그래도 null일 수 있다는 가능성을 고려하기만 했는데도 어떤 상황에 얘가 null일 수 있고,
그럼 어떻게 처리 하는게 좋은지 고민해보게 되네요,
나중에 수정해보도록 하겠습니다,,!

감사합니다

yujo11 commented on src/components/MovieRenderer.ts:
```diff
@@ -1,63 +1,78 @@
+  movieGrid.insertAdjacentHTML("beforeend", generateSkeletons(8));
+};
+
+export const removeSkeletonUI = (movieContainer: HTMLElement): void => {
+  movieContainer
+    .querySelectorAll(".movie-item.skeleton")
+    .forEach((el) => el.remove());
+};
```
부모 요소에 속성을 줘서 스타일이 적용되게 하면 하나씩 요소를 순회하며 찾아가는 대신 한번에 없애버릴 수도 있을거 같아요

ex)
```
<parents class="skeleton">
  <movie-item>
</parents>

```

devpearlkim replied:
감사합니다,,!

yujo11 reviewed (changes requested):
안녕하세요 진주님 ~

코드의 추가와 변경도 많았는데 그 와중에 UX를 고려해서 추가로 별점에 mouse hover시 효과를 구현하신게 인상적이었네요!

코드를 보면서 궁금한 점들과 소소하게 개선이 가능할거 같은 부분들 코멘트로 남겼는데 확인해주시면 좋을거 같습니다! 고생 많으셨습니다 진주님 ~~ 👍 👍 👍

devpearlkim committed:
fix: 피드백 반영 수정

- Refactored DOM selection using class-based styling instead of deep querySelector traversal.
- Moved domain-specific logic from utils/storage to services/userRating.
- Updated hasMore type in the infinite scroll hook.
- Applied early returns for cleaner logic flow.
- Replaced null assertions (!) with explicit null checks in Modal.
- Wrapped skeleton elements for more efficient removal.

devpearlkim committed:
fix: use Element.closest() for star click events

devpearlkim commented:
감사합니다 윤호님, 
지금 다시 확인해 보니,
로컬스토리지에 별점 저장하는것도 잘 작동을 안하고
UX 적으로도 수정할 부분이 많은 것 같습니다.

말씀해주신 피드백들은 나중에 다시 보면서 앞으로도 일하면서도 잘 참고하겠습니다.
피드백 해주신 내용들 보며 무심코 넘겼던 부분들에 대해 생각해볼 수 있었습니다.
감사합니다

yujo11 commented on index.html:
```diff
@@ -51,7 +51,7 @@ <h1 class="logo">
         <main>
           <section>
-            <h2></h2>
+            <h2 class='section-title'></h2>
```
👍 👍 👍

yujo11 reviewed (approved):
고생 많으셨습니다 진주님~ 이전 코멘트에 추가로 남긴 내용이 있으니 확인해주시면 좋을거 같아요!

이제 마지막 step만 남았는데요. step1 때 말씀 드렸던 것처럼 class로 구조를 변경해보시는 것도 좋은 챌린지가 될거 같네요! 마지막까지 화이팅입니다 ~~ 👍 👍 👍
